nncf/torch/nested_objects_traversal.py

```diff
@@ -0,0 +1,161 @@
+from functools import partial
+from typing import Callable, Mapping, Sequence, Set, List, Type, Any, Dict
+from typing import Tuple
+from typing import Union
+
+
+string_types = (str, bytes)
+iteritems = lambda mapping: getattr(mapping, 'iteritems', mapping.items)()
+
+
+def to_tuple(lst: List,
+             named_tuple_class: Type = None,
+             named_tuple_fields: List[str] = None) -> Tuple:
+    # Able to produce namedtuples if a corresponding parameter is given
+    if named_tuple_fields is None:
+        return tuple(lst)
+    return named_tuple_class(*lst)
+
+
+def is_tuple(obj) -> bool:
+    return isinstance(obj, tuple)
+
+
+def is_named_tuple(obj) -> bool:
+    return is_tuple(obj) and (obj.__class__ != tuple)
+
+
+def maybe_get_iterator(obj):
+    it = None
+        # pylint:disable=isinstance-second-argument-not-valid-type
+    if isinstance(obj, Mapping):
+        it = iteritems
+        # pylint:disable=isinstance-second-argument-not-valid-type
+    elif isinstance(obj, (Sequence, Set)) and not isinstance(obj, string_types):
+        it = enumerate
+    return it
+
+
+class InputIndexEntry:
+    def __init__(self, path: Tuple[Union[int, str], ...], getter: Callable, setter: Callable):
+        self.path = path
+        self.getter = getter
+        self.setter = setter
+
+
+class TupleRebuildingSetter:
+    def __init__(self, idx_to_set, current_tuple, previous_level_setter_for_current_tuple):
+        self._previous_level_setter = previous_level_setter_for_current_tuple
+        self._current_tuple = current_tuple
+        self._idx_to_set = idx_to_set
+
+    def __call__(self, value):
+        tmp_list = list(self._current_tuple)
+        tmp_list[self._idx_to_set] = value
+        new_tuple = tuple(tmp_list)
+        self._current_tuple = new_tuple
+        self._previous_level_setter(new_tuple)
+
+
+def nested_object_paths_generator(obj, out_entries_list, path=(), memo=None, previous_level_setter=None):
```

Comment:
This should not be a free function, and a function with side effects. If you need to reuse the path generation functionality  in more places than one, then encapsulate it into a proper class object with well-defined inputs and outputs and no side effects. Use the object wherever needed.

You can call the new class `NestedObjectIndex`, storing and accessing `IndexEntry` objects like it was done before for inputs, and reuse it, correspondingly, wherever the flat nested object indexing is required, such as a) for input wrapping and input port detection, b) for differential output deduction, and c) (in the future) for output port detection.

Reply:
Great job.